File: Maximum_subarray_sum/Kadane's_algorithm/kadane.cpp
#include <iostream>
#include <vector>
#include <algorithm>


int kadane(std::vector<int>& val)
{

	int max_sum = val[0];
	int curr_sum = 0;

	for (auto& ele : val)
	{

		curr_sum = std::max(curr_sum + ele, ele);
		max_sum = std::max(curr_sum, max_sum);

	}

	return (max_sum);

}


int main()
{

	std::vector<int> values;

	std::cout << "Enter the elements and terminate the input with any character [A to Z]:\n";

	for (int element; std::cin >> element;)
		values.push_back(element);

	if (values.size() == 0)
	{

		std::cout << "Empty array." <<std::endl;
		exit(1);

	}

	int max_sum = kadane(values);         // Calling the Kadane's algorithm. It will return the value of the maximum subarray sum.

	std::cout << "\nThe largest sum of subarray is: " << max_sum << std::endl;

	return 0;
}

File: Maximum_subarray_sum/Kadane's_algorithm/kadane.py
def kadane(val):

    prev_sum = 0
    
    try:
        
        max_sum = val[0]
    
    except IndexError:
        
        return "Array is empty."
        
    for x in val:
    
        prev_sum = max(prev_sum + x, x)

        max_sum = max(max_sum, prev_sum)        


    return max_sum

###########################

values = list(input("Enter the elements:\n").split())

values = [int(x) for x in values]

max_sum = kadane(values)

print("\n", max_sum, sep="", end="\n")
